vimeo/VimeoNetworking
# VIM-7242: Networking refactor

#### Ticket

[VIM-7242](https://vimean.atlassian.net/browse/VIM-7242)

#### Pull Request Checklist


#### Issue Summary

Currently we have multiple targets that depend directly on `AFNetworking` for various reasons. These include `VimeoKit`, `VimeoUpload`, `iOS` and `macOS` (to a lesser extent).

The goal of this PR is make the necessary changes to allow us to prevent these targets from consuming `AFNetworking` directly, and only exposing the minimum functionality required via `VimeoNetworking` as a proxy.

#### Implementation Summary

This PR has been broken down into a few different subtickets that have been previously reviewed by various members of the team. The combined changes in this feature branch include the following:

* Introduction of a new `NetworkReachabilityManager` that mirrors the API and functionality previosuly provided by `AFNetworkingReachabilityManager`. This also includes various unit testing classes.

* Small refactor of `VimeoClient` and `VimeoSessionManager` to move some of the request logic to live inside `VimeoSessionManager`.

* New `SessionManaging` & `AuthenticationListeningDelegate` protocols to abstract `VimeoClient`'s direct dependency on `VimeoSessionManager`

* Addition of various new protocols to help abstract our dependency on `AFNetworking` - `EndpointType`, `Cancellable` and `URLRequestConvertible`.

* Addition of new URL and JSON request serializers (currently unused), both thoroughly tested.

#### Reviewer Tips

To help with context and motivation for some of these changes it might be helpful to review the PRs currently open on the frameworks that currently dependend on AFNetworking:

[VimeoKit](https://github.vimeows.com/MobileApps/VimeoKit/pull/473)
[VimeoUpload](https://github.com/vimeo/VimeoUpload/pull/230)
[Vimeo iOS](https://github.vimeows.com/MobileApps/Vimeo-iOS/pull/2972)

#### How to Test

- Build and run, ensure all tests pass
- Switch to the iOS repo and checkout the branch on the PR above, ensure there are no changes in the current app behavior and that all tests pass. In particular please make sure you can upload videos as usual.
